spmd/compiler/fusion.py

```diff
@@ -0,0 +1,182 @@
+    size: int = 0
+    prev_node: Optional[fx.Node] = None  # node that was before start of section
+    next_node: Optional[fx.Node] = None  # node that was after end of section
+    processed: bool = False
+    output_name: str = ""
+    clone_node: Optional[fx.Node] = None
+    comm_node: Optional[fx.Node] = None
+    wait_node: Optional[fx.Node] = None
+
+
+@dataclass
+class GraphInfo:
+    len: int = 0
+    global_buffer: Optional[fx.Node] = None
+    global_buffer_size: int = 0
+    output: Optional[fx.Node] = None
+    first: Optional[fx.Node] = None
+
+    def update_info(self, gm: fx.GraphModule, debug: bool = True) -> None:
+        """get the len, input and output nodes"""
+        graph_len = gm.graph._len
+        if not graph_len:
+            raise ValueError("Empty graph passed in....")
+        self.len = graph_len
+
+        nodelist = gm.graph.nodes
+
+        for i, node in enumerate(nodelist):
+            if node.op == OP.PLACEHOLDER:
+                self.first = node
+                break
+
+        self.output = get_output_node(gm)
+        assert (
+            self.output is not None
+        ), f"unable to locate output node in gm {gm.graph}"
+
+        if debug:
+            print(
+                f"updated info - len = {self.len} input = {self.first}, output = {self.output}"
+            )
+
+
+def _insert_buffer_node(
+    gm: fx.GraphModule, insert_before_node: fx.Node, buffer_size: Iterable[int]
+) -> fx.Node:
+    """insert a torch.empty node in front of insert_before_node"""
+    with gm.graph.inserting_before(insert_before_node):
+        new_buffer_node = gm.graph.create_node(
+            OP.CALL_FUNCTION,
+            target=torch.empty,
+            args=tuple(buffer_size),  # TODO - need device
+        )
+    assert (
+        new_buffer_node is not None
+    ), f"failed to create buffer node, size={buffer_size}"
+
+    return new_buffer_node
+
+
+def _scan_graph_for_fusion_elements(
+    gm: fx.GraphModule,
+    comm_type: Comm_Type = Comm_Type.allreduce,
+) -> Optional[List[FusionElement]]:
+    """scan entire graph for matching sections of CommTensor style expansions
+    returns list of FusionElements that match comm_type"""
+
+    element_list = []
+
+    fe_pattern = [
+        "clone",
+        "_tensor_constant",
+        "_tensor_constant",
+        comm_type,
+        "comm_result",
+        "getitem",
+        "getitem",
+        "wait_comm",
+    ]
+
+    fe_size = len(fe_pattern) - 1
+    curr_count = 0
+    curr_node_list = []
+    rank = dist.get_rank()
+
+    for i, node in enumerate(gm.graph.nodes):
+
+        pattern = fe_pattern[curr_count]
+
+        if curr_count < fe_size:
+
+            if node.name.startswith(pattern):
+                curr_node_list.append(node)
+                curr_count += 1
+                continue
+            else:
+                curr_count == 0
+                curr_node_list.clear()
+                continue
+
+        if curr_count == fe_size:
```

Comment:
This can be `elif` which will make the code more consistent.

Reply:
great point.